### ISSUE TITLE
TorchScript Poor CUDA Performance 

### ISSUE THREAD

**Issue by madsbk** [pytorch/pytorch#23655]
The performance of a TorchScript that handles the computation in the [Adam Optimizer](https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py) is worse than using the native PyTorch operations on CUDA. 

### Background
I am trying to improve the GPU performance of the [Adam Optimizer](https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py) since it is the bottleneck in one of our workflows. 

### Results
I have written a standalone script that compares PyTorch native tensor operations, TorchScript, and Numba (three runs each):
```
native :  0.0506
native :  0.0322
native :  0.0322
TScript:  0.1917
TScript:  0.0442
TScript:  0.0442
numba  :  0.1638
numba  :  0.0145
numba  :  0.0143
```
I suspect the problem is that TorchScript does not fuse all operations into a single CUDA. It generates two CUDA kernels (see the `PYTORCH_FUSION_DEBUG` output below)

### Standalone Script
```python
import math
import torch
import time
import random
import sys

def kernel_native(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    # type: (Tensor, Tensor, Tensor, Tensor, float, float, float, float) -> None
    exp_avg.mul_(beta1).add_(1 - beta1, grad)
    exp_avg_sq.mul_(beta2).addcmul_(1 - beta2, grad, grad)
    denom = exp_avg_sq.sqrt().add_(eps)
    param.addcdiv_(-step_size, exp_avg, denom)   

@torch.jit.script
def kernel_jit_script(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    # type: (Tensor, Tensor, Tensor, Tensor, float, float, float, float) -> None
    exp_avg.mul_(beta1)
    exp_avg.add_((1 - beta1) + grad)
    exp_avg_sq.mul_(beta2)
    exp_avg_sq.add_((1 - beta2)*(grad*grad))
    denom = exp_avg_sq.sqrt().add_(eps)
    param += -step_size * (exp_avg / denom)

from numba import cuda
@cuda.jit
def kernel_numba_cuda(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    i = cuda.grid(1)
    if i > grad.size:
        return

    exp_avg[i] = exp_avg[i] * beta1 + (1 - beta1) + grad[i]
    exp_avg_sq[i] = exp_avg_sq[i] * beta2 + (1 - beta2) *grad[i]*grad[i]
    denom = math.sqrt(exp_avg_sq[i]) + eps
    param[i] += -step_size * (exp_avg[i] / denom)

def kernel_numba(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    # type: (Tensor, Tensor, Tensor, Tensor, float, float, float, float) -> None

    import ctypes
    import numpy as np
    def get_devicendarray(t):
        assert t.type() == 'torch.cuda.FloatTensor'
        numpy_dtype = np.float32
        itemsize = 4

        ctx = cuda.cudadrv.driver.driver.get_context()
        mp = cuda.cudadrv.driver.MemoryPointer(ctx, ctypes.c_ulong(t.data_ptr()), t.numel()*itemsize)
        return cuda.cudadrv.devicearray.DeviceNDArray(t.numel(), itemsize, numpy_dtype(), gpu_data=mp, stream=torch.cuda.current_stream().cuda_stream)

    numba_param = get_devicendarray(param)
    numba_grad = get_devicendarray(grad)
    numba_exp_avg = get_devicendarray(exp_avg)
    numba_exp_avg_sq = get_devicendarray(exp_avg_sq)
    threadsperblock = 32
    blockspergrid = (param.numel() + (threadsperblock - 1)) // threadsperblock
    kernel_numba_cuda[blockspergrid, threadsperblock](numba_param, numba_grad, numba_exp_avg, numba_exp_avg_sq, beta1, beta2, step_size, eps)

kernels = [("native ", kernel_native), ("TScript", kernel_jit_script), ("numba  ", kernel_numba)]
shape = (4194304*64,)
param, grad, exp_avg, exp_avg_sq = [torch.rand(shape).to("cuda") for x in range(4)]
beta1, beta2, step_size, eps = [random.random() for x in range(4)]
torch.cuda.synchronize()

for name, func in kernels:
    for i in range(3):
        t1 = time.time()
        func(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps)
        torch.cuda.synchronize()
        t2 = time.time()
        print("%s: " % name, t2-t1)
```

### Output of PYTORCH_FUSION_DEBUG
```cu
fusion code:

typedef unsigned char uint8_t;
typedef signed char int8_t;
typedef short int  int16_t;
typedef long long int int64_t;



#define NAN __int_as_float(0x7fffffff)
#define POS_INFINITY __int_as_float(0x7f800000)
#define NEG_INFINITY __int_as_float(0xff800000)

typedef unsigned int IndexType;
template<typename T, size_t N>
struct TensorInfo {
  T* data;
  IndexType sizes[N];
  IndexType strides[N];
};
template<typename T>
struct TensorInfo<T, 0> {
  T * data;
};


extern "C" __global__
void kernel_0(IndexType totalElements, const TensorInfo<float,1> t0, double s1, const TensorInfo<float,1> t2 ) {
  
  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
        linearIndex < totalElements;
        linearIndex += gridDim.x * blockDim.x) {
      // Convert `linearIndex` into an offset of tensor:
      IndexType t0_offset = 0;
      IndexType t0_linearIndex = linearIndex;
      
      //printf("tensor t0 sizes[0] = %d, strides[0] = %d\n", t0.sizes[0],t0.strides[0]);
      size_t t0_dimIndex0 = t0_linearIndex ;
      t0_offset += t0_dimIndex0 ;
      IndexType t2_offset = 0;
      IndexType t2_linearIndex = linearIndex;
      
      //printf("tensor t2 sizes[0] = %d, strides[0] = %d\n", t2.sizes[0],t2.strides[0]);
      size_t t2_dimIndex0 = t2_linearIndex ;
      t2_offset += t2_dimIndex0 ;
      
      // calculate the results
      float n0 = __ldg(&t0.data[t0_offset]);
      double n1 = s1;
      float n2 = n0 * n0;
      float n3 = n2 * ((float) n1);
      t2.data[t2_offset] = n3;
      
    }
}

fusion code:

typedef unsigned char uint8_t;
typedef signed char int8_t;
typedef short int  int16_t;
typedef long long int int64_t;



#define NAN __int_as_float(0x7fffffff)
#define POS_INFINITY __int_as_float(0x7f800000)
#define NEG_INFINITY __int_as_float(0xff800000)

typedef unsigned int IndexType;
template<typename T, size_t N>
struct TensorInfo {
  T* data;
  IndexType sizes[N];
  IndexType strides[N];
};
template<typename T>
struct TensorInfo<T, 0> {
  T * data;
};


extern "C" __global__
void kernel_1(IndexType totalElements, const TensorInfo<float,1> t0, const TensorInfo<float,1> t1, double s2, const TensorInfo<float,1> t3 ) {
  
  for (IndexType linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
        linearIndex < totalElements;
        linearIndex += gridDim.x * blockDim.x) {
      // Convert `linearIndex` into an offset of tensor:
      IndexType t0_offset = 0;
      IndexType t0_linearIndex = linearIndex;
      
      //printf("tensor t0 sizes[0] = %d, strides[0] = %d\n", t0.sizes[0],t0.strides[0]);
      size_t t0_dimIndex0 = t0_linearIndex ;
      t0_offset += t0_dimIndex0 ;
      IndexType t1_offset = 0;
      IndexType t1_linearIndex = linearIndex;
      
      //printf("tensor t1 sizes[0] = %d, strides[0] = %d\n", t1.sizes[0],t1.strides[0]);
      size_t t1_dimIndex0 = t1_linearIndex ;
      t1_offset += t1_dimIndex0 ;
      IndexType t3_offset = 0;
      IndexType t3_linearIndex = linearIndex;
      
      //printf("tensor t3 sizes[0] = %d, strides[0] = %d\n", t3.sizes[0],t3.strides[0]);
      size_t t3_dimIndex0 = t3_linearIndex ;
      t3_offset += t3_dimIndex0 ;
      
      // calculate the results
      float n0 = __ldg(&t0.data[t0_offset]);
      float n1 = __ldg(&t1.data[t1_offset]);
      double n2 = s2;
      float n3 = n0 / n1;
      float n4 = n3 * ((float) n2);
      t3.data[t3_offset] = n4;
      
    }
}
```

**Comment by fmassa** [comment]
@madsbk can you try removing all the in-place operations in your scripted version?
IIRC in-place ops block fusion.

I was btw able to fuse most of sparse adam by removing those in-place ops, see https://github.com/pytorch/pytorch/pull/23522#issuecomment-516952462

**Comment by soumith** [comment]
cc: @wanchaol and also relevant WIP PR: [JIT version of Adagrad](https://github.com/pytorch/pytorch/pull/23640)

**Comment by wanchaol** [comment]
Yeah the CUDA fusion does not support fusing in-place operations yet, Can you try the suggestion by @fmassa to see if you get a better perf?

**Comment by wanchaol** [comment]
@madsbk Also, it seems like you are only timing it in 3 iterations, and the perf numbers seems including the compilation time. The first time JIT/TorchScript sees an input it will specialize and do optimizations against that input, so when comparing performance please allow a compilation time and runtime optimization warmup for the input.

**Comment by xsacha** [comment]
@wanchaol That made a big difference to my traced module too. Any other tricks like this? It slows down the non-JIT version, however.

**Comment by wanchaol** [comment]
@xsacha That's pretty much all, let me summarize:

1. JIT fusion does not support in-place operation fusion and reduction operation yet.
2. we did the fusion code generation at the first time we see an input that the shape we haven't seen before (like if the number of dimensions are different), this will help us generate faster cuda kernels. 

So the TorchScript compilation time will be gone after compile and the first run :)

**Comment by madsbk** [comment]
Thanks for the feedback!

I have removed all in-place operations and added three kernel implementations:
```python

import math
import torch
import time
import random
import sys

def kernel_native(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    # type: (Tensor, Tensor, Tensor, Tensor, float, float, float, float) -> None
    exp_avg.mul_(beta1).add_(1 - beta1, grad)
    exp_avg_sq.mul_(beta2).addcmul_(1 - beta2, grad, grad)
    denom = exp_avg_sq.sqrt().add_(eps)
    param.addcdiv_(-step_size, exp_avg, denom)   

@torch.jit.script
def kernel_jit_script1(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    # type: (Tensor, Tensor, Tensor, Tensor, float, float, float, float) -> None
    exp_avg *= beta1
    exp_avg += (1 - beta1) + grad
    exp_avg_sq *= beta2
    exp_avg_sq += (1 - beta2)*(grad*grad)
    param += -step_size * (exp_avg / exp_avg_sq.sqrt() + eps)

@torch.jit.script
def kernel_jit_script2(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    # type: (Tensor, Tensor, Tensor, Tensor, float, float, float, float) -> None
    exp_avg[:] = exp_avg * beta1 + (1 - beta1) + grad
    exp_avg_sq[:] = exp_avg_sq * beta2 + (1 - beta2)*(grad*grad)
    param[:] = param -step_size * (exp_avg / exp_avg_sq.sqrt() + eps)

@torch.jit.script
def kernel_jit_script3(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    # type: (Tensor, Tensor, Tensor, Tensor, float, float, float, float) -> Tuple[Tensor, Tensor, Tensor, Tensor]
    exp_avg = exp_avg * beta1 + (1 - beta1) + grad
    exp_avg_sq = exp_avg_sq * beta2 + (1 - beta2)*(grad*grad)
    param = param -step_size * (exp_avg / exp_avg_sq.sqrt() + eps)
    return (param, grad, exp_avg, exp_avg_sq)


from numba import cuda
@cuda.jit
def kernel_numba_cuda(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    i = cuda.grid(1)
    if i > grad.size:
        return

    exp_avg[i] = exp_avg[i] * beta1 + (1 - beta1) + grad[i]
    exp_avg_sq[i] = exp_avg_sq[i] * beta2 + (1 - beta2) *grad[i]*grad[i]
    denom = math.sqrt(exp_avg_sq[i]) + eps
    param[i] += -step_size * (exp_avg[i] / denom)

def kernel_numba(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps):
    # type: (Tensor, Tensor, Tensor, Tensor, float, float, float, float) -> None

    import ctypes
    import numpy as np
    def get_devicendarray(t):
        assert t.type() == 'torch.cuda.FloatTensor'
        numpy_dtype = np.float32
        itemsize = 4

        ctx = cuda.cudadrv.driver.driver.get_context()
        mp = cuda.cudadrv.driver.MemoryPointer(ctx, ctypes.c_ulong(t.data_ptr()), t.numel()*itemsize)
        return cuda.cudadrv.devicearray.DeviceNDArray(t.numel(), itemsize, numpy_dtype(), gpu_data=mp, stream=torch.cuda.current_stream().cuda_stream)

    numba_param = get_devicendarray(param)
    numba_grad = get_devicendarray(grad)
    numba_exp_avg = get_devicendarray(exp_avg)
    numba_exp_avg_sq = get_devicendarray(exp_avg_sq)
    threadsperblock = 32
    blockspergrid = (param.numel() + (threadsperblock - 1)) // threadsperblock
    kernel_numba_cuda[blockspergrid, threadsperblock](numba_param, numba_grad, numba_exp_avg, numba_exp_avg_sq, beta1, beta2, step_size, eps)

kernels = [("native ", kernel_native), ("numba  ", kernel_numba)]
shape = (4194304*64,)
param, grad, exp_avg, exp_avg_sq = [torch.rand(shape).to("cuda") for x in range(4)]
beta1, beta2, step_size, eps = [random.random() for x in range(4)]
torch.cuda.synchronize()

kernels += [("TScript1", kernel_jit_script1), ("TScript2", kernel_jit_script2), ("TScript3", kernel_jit_script3)]
runs = 1

for name, func in kernels:
    func(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps)
    torch.cuda.synchronize()
    for i in range(runs):
        t1 = time.time()
        res = func(param, grad, exp_avg, exp_avg_sq, beta1, beta2, step_size, eps)
        if res is not None:
            param[:], _, exp_avg[:], exp_avg_sq[:] = res
        torch.cuda.synchronize()
        t2 = time.time()
        print("%s: " % name, t2-t1)
```
The results are:
```
native :  0.0323
numba  :  0.0147
TScript1:  0.0369
TScript2:  0.0352
TScript3:  0.0280
```
TScript3 is now better than native but still far from numba. TScript3 is now generating a single CUDA kernel but the back-copying of the result now dominates the timing. If I remove the line `param[:], _, exp_avg[:], exp_avg_sq[:] = res` the performance is almost on pair with numba:
```
TScript3 (no back-copy):  0.01664
```
The problem is I really do need to update the values :/

**Comment by eellison** [comment]
cc @mruberry

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
